sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java

```diff
@@ -0,0 +1,387 @@
+          OrderByFieldPath fieldPath =
+              OrderByFieldPath.fromString(filter.getUnaryFilter().getField().getFieldPath());
+          if (!result.contains(fieldPath)) {
+            result.add(fieldPath);
+          }
+        }
+        break;
+      default:
+        break;
+    }
+  }
+
+  static @Nullable Value lookupDocumentValue(Document document, String fieldPath) {
+    OrderByFieldPath resolvedPath = OrderByFieldPath.fromString(fieldPath);
+    // __name__ is a special field and doesn't exist in (top-level) valueMap
+    if (resolvedPath.isDocumentName()) {
+      return Value.newBuilder().setReferenceValue(document.getName()).build();
+    }
+    return findMapValue(new ArrayList<>(resolvedPath.getSegments()), document.getFieldsMap());
+  }
+
+  private static @Nullable Value findMapValue(List<String> segments, Map<String, Value> valueMap) {
+    if (segments.isEmpty()) {
+      return null;
+    }
+    String field = segments.remove(0);
+    Value value = valueMap.get(field);
+    if (segments.isEmpty()) {
+      return value;
+    }
+    // Field path traversal is not done, recurse into map values.
+    if (value == null || !value.getValueTypeCase().equals(ValueTypeCase.MAP_VALUE)) {
+      return null;
+    }
+    return findMapValue(segments, value.getMapValue().getFieldsMap());
+  }
+
+  private static class OrderByFieldPath implements Comparable<OrderByFieldPath> {
+
+    public static OrderByFieldPath fromString(String fieldPath) {
+      if (fieldPath.isEmpty()) {
+        throw new IllegalArgumentException("Could not resolve empty field path");
+      }
+      List<String> segments = new ArrayList<>();
+      while (!fieldPath.isEmpty()) {
+        Matcher segmentMatcher = FIELD_PATH_SEGMENT_REGEX.matcher(fieldPath);
+        boolean foundMatch = segmentMatcher.lookingAt();
+        if (!foundMatch) {
+          throw new IllegalArgumentException("OrderBy field path was malformed");
+        }
+        String fieldName;
+        if ((fieldName = segmentMatcher.group(1)) != null) {
+          segments.add(fieldName);
+        } else if ((fieldName = segmentMatcher.group(2)) != null) {
+          String unescaped = unescapeFieldName(fieldName.substring(1, fieldName.length() - 1));
+          segments.add(unescaped);
+        } else {
+          throw new IllegalArgumentException("OrderBy field path was malformed");
+        }
+        fieldPath = fieldPath.substring(fieldName.length());
+        // Due to the regex, any non-empty fieldPath will have a dot before the next nested field.
+        if (fieldPath.startsWith(".")) {
+          fieldPath = fieldPath.substring(1);
+        }
+      }
+      return new OrderByFieldPath(ImmutableList.copyOf(segments));
+    }
+
+    private final ImmutableList<String> segments;
+
+    private OrderByFieldPath(ImmutableList<String> segments) {
+      this.segments = segments;
+    }
+
+    public String canonicalString() {
+      return String.join(".", segments);
```

Comment:
The segments must be escaped for this to be valid.

Rather than also writing the code for escaping, one suggestion would be to save the escaped field path this instance was constructed from, and return it. It will not be the canonical escaped representation, but it will be valid.

A unit test should show the issue. For ex a query with no order by and an inequality filter on `foo.bar`.bar would add foo.bar.baz to the order by, which is different.

Reply:
Ah yeah, added a test case that shows it. I've switched it to returning the original string.